server/model.go

```diff
@@ -0,0 +1,254 @@
+package server
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/ollama/ollama/api"
+	"github.com/ollama/ollama/convert"
+	"github.com/ollama/ollama/llm"
+	"github.com/ollama/ollama/types/model"
+	"github.com/ollama/ollama/types/ordered"
+)
+
+func parseFromModel(ctx context.Context, name model.Name, fn func(api.ProgressResponse)) (*ordered.Map[*Layer, *llm.GGML], error) {
+	modelpath := ParseModelPath(name.DisplayLongest())
+	manifest, _, err := GetManifest(modelpath)
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		if err := PullModel(ctx, name.DisplayLongest(), &registryOptions{}, fn); err != nil {
+			return nil, err
+		}
+
+		return parseFromModel(ctx, name, fn)
```

Comment:
Can we just set `modelpath` and `manifest` here then continue? I prefer to avoid recursion when it's not needed

Reply:
done